File: practical 3/binary_search.cpp
#include <iostream>
#include <vector>
using namespace std;

bool binary_search(vector<int> vec, int target) {
    size_t lower_bound = 0;
    size_t upper_bound = vec.size();
    size_t mid = upper_bound / 2;
    while (lower_bound < upper_bound) {
        if (target == vec[mid]) {
            return true;
        } else if (target < vec[mid]) {
            upper_bound = mid;
        } else {
            lower_bound = mid;
        }
        mid = (upper_bound + lower_bound) / 2;
    }
    return false;
};
int main() {
    vector<int> oreo = {1,2,3,4,5,6,7,8,9,10};
    cout << binary_search(oreo, -1);
}

File: practical 3/binary_search.rs
fn binary_search(vec : &Vec<i32>, target : i32) -> bool {
    let mut upper_bound = vec.len();
    let mut lower_bound = 0;
    loop {
        let mid = (upper_bound + lower_bound)/2;

        if lower_bound >= upper_bound {
            return false;
        } else if vec[mid] == target {
            return true;
        } else if vec[mid] < target {
            lower_bound = mid + 1;
        } else if vec[mid] > target {
            upper_bound = mid - 1;
        }
    };
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn normal_search_if_target_exists_in_vec() {
        let vec = vec![1,22,32,43,44,66,77,100];
        assert!(binary_search(&vec, 22));
        assert!(binary_search(&vec, 32));
        assert!(binary_search(&vec, 100));
    }
    
    #[test]
    #[should_panic]
    fn search_if_target_doesnt_exist_in_vec() {
        let vec = vec![1,22,32,43,44,66,77,100];
        assert!(binary_search(&vec, 21));
        assert!(binary_search(&vec, 31));
        assert!(binary_search(&vec, 110));
    }


}